src/Components/CommandPrompt/CommandPrompt.tsx

```diff
@@ -0,0 +1,175 @@
+import Symbol from "Components/Symbol/Symbol";
+import { COMMAND_LIST } from "../../commands";
+import { Command } from "Models/Command";
+import Overlay from "Components/Overlay/Overlay";
+
+function CommandPrompt() {
+    const inputField = React.createRef<HTMLInputElement>();
+
+    const [shown, setShown] = useState<boolean>(false);
+    const [typed, setTyped] = useState("");
+    const [selectedCommandIndex, setSelectedCommandIndex] = useState(null);
+    const [matchedCommands, setMatchedCommands] = useState(COMMAND_LIST);
+    const commandToRun = useRef<Command>(null);
+
+
+    useEffect(() => {
+        setMatchedCommands(COMMAND_LIST.filter((command) => {
+            const lowerCommand = command.name.toLowerCase();
+            const lowerTyped = typed.toLowerCase();
+            return lowerCommand.includes(lowerTyped);
+        }))
+        setSelectedCommandIndex(matchedCommands.length >= 0 ? 0 : -1);
+    }, [typed])
+
+    useEffect(() => {
+        if (shown) {
+            inputField.current.focus();
+            setSelectedCommandIndex(0);
+        } else {
+            inputField.current.blur();
+            setTyped("");
+            if (commandToRun.current) {
+                commandToRun.current.callback();
+                commandToRun.current = null;
+            }
+        }
+    }, [shown])
+
+    useEffect(() => {
+        document.addEventListener('keydown', handleKeyDownInDocument, { capture: false });
+
+        return function cleanup () {
+            document.removeEventListener('keydown', handleKeyDownInDocument, { capture: false });
+        }
+    }, [shown, selectedCommandIndex])
+
+    const handleKeyDownInDocument = (event) => {
+        // TODO: Shortcut in a settings pane ?
+        if (event.ctrlKey && event.key == 'k') {
+            setShown(() => !shown);
+            event.preventDefault();
+        }
+
+        if (shown && event.key == "Escape") {
+            setShown(false);
+            event.preventDefault();
+        }
+
+        // Selected item is cyclic
+        switch (event.key) {
+            case "ArrowUp":
+                setSelectedCommandIndex((selectedCommandIndex + 1) % matchedCommands.length);
+                event.preventDefault();
+                break
+            case "ArrowDown":
+                setSelectedCommandIndex((selectedCommandIndex - 1 + matchedCommands.length) % matchedCommands.length);
+                event.preventDefault();
+                break
+        }
+    }
+
+    const handleValueChangeInInput = (event) => setTyped(event.target.value);
+
+    const handleInputKeyDown = (event) => {
+        if (event.key == "Enter") {
+            runSelectedCommand(matchedCommands.length != 0 ? selectedCommandIndex : null);
+        }
+    }
+
+    const runSelectedCommand = (index: number | null) => {
+        if (index != null) {
+            commandToRun.current = matchedCommands[index];
+            setShown(false);
+        }
+    }
```

Comment:
Maybe use undefined here and a guard statement:

```suggestion
    const runSelectedCommand = (index?: number) => {
        if (!index) return;

        commandToRun.current = matchedCommands[index];
        setShown(false);
    }
```